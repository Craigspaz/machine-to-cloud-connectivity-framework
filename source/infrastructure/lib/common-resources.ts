// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement, ServicePrincipal } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, CfnBucket } from '@aws-cdk/aws-s3';

/**
 * Machine to Cloud Connectivity Framework Common Resources
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  private cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  private s3LoggingBucket: Bucket;
  // S3 Resource Storage bucket
  private s3Bucket: Bucket;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });
    const cfnBucket = this.s3LoggingBucket.node.defaultChild as CfnBucket;
    cfnBucket.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
        { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
      ]
    });

    this.s3Bucket = new Bucket(this, 'M2C2Bucket', {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: this.s3LoggingBucket,
      serverAccessLogsPrefix: 'm2c2/'
    });
    this.s3Bucket.addToResourcePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        principals: [ new ServicePrincipal('lambda.amazonaws.com') ],
        actions: [
          's3:GetObject',
          's3:PutObject'
        ],
        resources: [
          `${this.s3Bucket.bucketArn}/*`
        ]
      })
    );
  }

  /**
   * Get CloudWatch Logs policy.
   * @return {Policy} CloudWatch Logs policy
   */
  getCloudWatchLogsPolicy(): Policy {
    return this.cloudWatchLogsPolicy;
  }

  /**
   * Get S3 logging bucket.
   * @return {Bucket} Logging bucket
   */
  getS3LoggingBucket(): Bucket {
    return this.s3LoggingBucket;
  }

  /**
   * Get S3 M2C2 bucket.
   * @return {Bucket} M2C2 bucket
   */
  getS3Bucket(): Bucket {
    return this.s3Bucket;
  }

  /**
   * Get S3 M2C2 bucket name
   * @return {string} M2C2 bucket name
   */
  getS3BucketName(): string {
    return this.s3Bucket.bucketName;
  }
}