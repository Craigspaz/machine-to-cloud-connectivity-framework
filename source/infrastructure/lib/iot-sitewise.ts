// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnDeletionPolicy, Construct } from '@aws-cdk/core';
import { CfnGateway } from '@aws-cdk/aws-iotsitewise';

/**
 * IoTSitewiseConstructProps props
 * @interface IoTSitewiseConstructProps
 */
export interface IoTSitewiseConstructProps {
  // Greengrass group ID
  readonly greengrassGroupId: string;
}

/**
 * @class
 * Machine to Cloud Connectivity Framework IoT Sitewise Construct.
 * It creates a common CloudWatch Logs policy for Lambda functions, a logging S3 bucket, and M2C2 S3 bucket.
 */
export class IoTSitewiseConstruct extends Construct {
  // IoT Sitewise gateway
  public iotSitewiseGateway: CfnGateway;

  constructor(scope: Construct, id: string, props: IoTSitewiseConstructProps) {
    super(scope, id);

    this.iotSitewiseGateway = new CfnGateway(this, 'Gateway', {
      gatewayName: `${Aws.STACK_NAME}-sitewise-gateway`,
      gatewayPlatform: {
        greengrass: {
          groupArn: `arn:${Aws.PARTITION}:greengrass:${Aws.REGION}:${Aws.ACCOUNT_ID}:/greengrass/groups/${props.greengrassGroupId}`
        }
      }
    });
    this.iotSitewiseGateway.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
  }
}