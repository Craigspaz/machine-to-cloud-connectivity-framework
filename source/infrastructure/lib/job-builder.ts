// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Aws,
  CfnDeletionPolicy,
  CfnMapping,
  Construct,
  Duration
} from '@aws-cdk/core';
import { AttributeType, CfnTable } from '@aws-cdk/aws-dynamodb';
import {
  CfnPolicy,
  Effect,
  Policy,
  PolicyStatement,
  Role,
  ServicePrincipal
} from '@aws-cdk/aws-iam';
import { CfnTopicRule } from '@aws-cdk/aws-iot';
import { Code, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';

/**
 * JobBuilderConstruct props
 */
export interface JobBuilderConstructProps {
  /**
   * Policy for CloudWatch Logs
   */
  readonly cloudWatchLogsPolicy: Policy;
  /**
   * Greengrass group ID
   */
  readonly greengrassGroupId: string;
  /**
   * Kinesis Data Stream name
   */
  readonly kinesisStreamName: string;
  /**
   * S3 M2C2 bucket
   */
  readonly s3Bucket: Bucket;
  /**
   * S3 logging bucket
   */
  readonly s3LoggingBucket: Bucket;
  /**
   * Sending anonymous usage
   */
  readonly sendAnonymousUsage: string;
  /**
   * Mapping for the source code
   */
  readonly sourceCodeMapping: CfnMapping;
  /**
   * Mapping for the solution information
   */
  readonly solutionMapping: CfnMapping;
  /**
   * Solutions UUID
   */
  readonly uuid: string;
}

/**
 * Machine to Cloud Connectivity Framework Job Builder Lambda, DynamoDB, S3, and IoT rule
 */
export class JobBuilderConstruct extends Construct {
  // Job metadata DynamoDB table name
  private dynamodbTableName: string;

  constructor(scope: Construct, id: string, props: JobBuilderConstructProps) {
    super(scope, id);

    const sourceCodeBucket: string = `${props.sourceCodeMapping.findInMap('General', 'S3Bucket')}-${Aws.REGION}`;
    const sourceCodePrefix: string = props.sourceCodeMapping.findInMap('General', 'KeyPrefix');

    const connectorLambdaRole = new Role(this, 'ConnectorLambdaRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path: '/'
    });
    connectorLambdaRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const lambdaToDynamoDb = new LambdaToDynamoDB(this, 'JobBuilder', {
      lambdaFunctionProps: {
        code: Code.fromBucket(
          Bucket.fromBucketName(this, 'SourceCodeBucket', sourceCodeBucket),
          `${sourceCodePrefix}/m2c2-job-builder.zip`
        ),
        description: 'Machine to Cloud Connectivity job builder function',
        environment: {
          CONNECTOR_LAMBDA_ROLE: connectorLambdaRole.roleArn,
          GREENGRASS_ID: props.greengrassGroupId,
          JOB_BUILDER_BUCKET: props.s3Bucket.bucketName,
          JOB_BUILDER_KEY: props.solutionMapping.findInMap('Parameters', 'JobMetadataPrefix'),
          KINESIS_STREAM: props.kinesisStreamName,
          SEND_ANONYMOUS_METRIC: props.sendAnonymousUsage,
          SOLUTION_ID: props.solutionMapping.findInMap('Parameters', 'Id'),
          SOLUTION_VERSION: props.solutionMapping.findInMap('Parameters', 'Version'),
          SOURCE_S3_BUCKET: sourceCodeBucket,
          SOURCE_S3_PREFIX: sourceCodePrefix,
          UUID: props.uuid
        },
        handler: 'm2c2_job_builder_main.lambda_handler',
        runtime: Runtime.PYTHON_3_8,
        timeout: Duration.minutes(5)
      },
      dynamoTableProps: {
        partitionKey: {
          name: 'jobid',
          type: AttributeType.STRING
        },
        sortKey: {
          name: 'version',
          type: AttributeType.NUMBER
        }
      },
      tableEnvironmentVariableName: 'JOB_DYNAMODB_TABLE'
    });

    const cfnDynamoTable = lambdaToDynamoDb.dynamoTable.node.defaultChild as CfnTable;
    cfnDynamoTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
    this.dynamodbTableName = lambdaToDynamoDb.dynamoTable.tableName;

    const jobBuilderPolicy = new Policy(this, 'JobBuilderPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [ 's3:GetObject' ],
          resources: [ `arn:${Aws.PARTITION}:s3:::${sourceCodeBucket}/${sourceCodePrefix}/*` ]
        }),
         new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            's3:GetBucketLocation',
            's3:GetObject',
            's3:ListBucket',
            's3:PutObject'
          ],
          resources: [
            props.s3Bucket.bucketArn,
            `${props.s3Bucket.bucketArn}/*`
          ]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [ 'iot:Publish' ],
          resources: [ `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/job/*` ]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'lambda:CreateFunction',
            'lambda:DeleteFunction',
            'lambda:PublishVersion',
            'lambda:CreateAlias',
            'lambda:DeleteAlias'
          ],
          resources: [
            `arn:${Aws.PARTITION}:lambda:${Aws.REGION}:${Aws.ACCOUNT_ID}:function:m2c2-opcda-connector*`,
            `arn:${Aws.PARTITION}:lambda:${Aws.REGION}:${Aws.ACCOUNT_ID}:function:m2c2-slmp-connector*`
          ]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [ 'iam:PassRole' ],
          resources: [ connectorLambdaRole.roleArn ]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'greengrass:CreateCoreDefinitionVersion',
            'greengrass:CreateCoreDefinition',
            'greengrass:CreateDeployment',
            'greengrass:CreateDeviceDefinitionVersion',
            'greengrass:CreateDeviceDefinition',
            'greengrass:CreateFunctionDefinitionVersion',
            'greengrass:CreateFunctionDefinition',
            'greengrass:CreateGroupVersion',
            'greengrass:CreateLoggerDefinitionVersion',
            'greengrass:CreateLoggerDefinition',
            'greengrass:CreateResourceDefinitionVersion',
            'greengrass:CreateResourceDefinition',
            'greengrass:CreateSubscriptionDefinitionVersion',
            'greengrass:CreateSubscriptionDefinition',
            'greengrass:DeleteCoreDefinition',
            'greengrass:DeleteDeviceDefinition',
            'greengrass:DeleteFunctionDefinition',
            'greengrass:DeleteLoggerDefinition',
            'greengrass:DeleteResourceDefinition',
            'greengrass:DeleteSubscriptionDefinition',
            'greengrass:GetCoreDefinitionVersion',
            'greengrass:GetDeploymentStatus',
            'greengrass:GetDeviceDefinitionVersion',
            'greengrass:GetFunctionDefinitionVersion',
            'greengrass:GetGroupVersion',
            'greengrass:GetGroup',
            'greengrass:GetLoggerDefinitionVersion',
            'greengrass:GetResourceDefinitionVersion',
            'greengrass:GetSubscriptionDefinitionVersion',
            'greengrass:ListCoreDefinitions',
            'greengrass:ListDeviceDefinitions',
            'greengrass:ListFunctionDefinitions',
            'greengrass:ListLoggerDefinitions',
            'greengrass:ListResourceDefinitions',
            'greengrass:ListSubscriptionDefinitions'
          ],
          resources: [
            '*'
          ]
        })
      ]
    });
    const cfnPolicy = jobBuilderPolicy.node.defaultChild as CfnPolicy;
    cfnPolicy.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W12', reason: 'The * resource is needed to control Greengrass resources.' }
      ]
    });

    const jobBuilderLambdaFunction = lambdaToDynamoDb.lambdaFunction;
    const jobBuilderRole = jobBuilderLambdaFunction.role as Role;
    jobBuilderRole.attachInlinePolicy(jobBuilderPolicy);

    const jobGeneratorRule = new CfnTopicRule(this, 'JobGeneratorRule', {
      topicRulePayload: {
        actions: [{
          lambda: { functionArn: jobBuilderLambdaFunction.functionArn }
        }],
        description: 'Processing of the job submitted and generating the job file to be pushed to Greengrass',
        ruleDisabled: false,
        sql: `SELECT * FROM 'm2c2/job/request'`
      }
    });

    jobBuilderLambdaFunction.addPermission('LambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('iot.amazonaws.com'),
      sourceAccount: Aws.ACCOUNT_ID,
      sourceArn: jobGeneratorRule.attrArn
    });
  }

  /**
   * Get DynamoDB table name.
   * @return {string} DynamoDB table name
   */
  getDynamoDbTableName(): string {
    return this.dynamodbTableName;
  }
}