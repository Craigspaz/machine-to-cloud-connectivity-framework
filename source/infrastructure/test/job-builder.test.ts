// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { CfnMapping, Stack } from '@aws-cdk/core';
import { Policy } from '@aws-cdk/aws-iam';
import { Bucket } from '@aws-cdk/aws-s3';
import { JobBuilderConstruct } from '../lib/job-builder';

test('M2C2 job builder test', () => {
  const stack = new Stack();
  const sourceCodeMapping = new CfnMapping(stack, 'SourceCode', {
    mapping: {
      General: {
        S3Bucket: 'Bucket',
        KeyPrefix: 'Prefix'
      }
    }
  });
  const solutionMapping = new CfnMapping(stack, 'Solution', {
    mapping: {
      Parameters: {
        Id: 'SolutionId',
        Version: 'SolutionVersion',
        JobMetadataPrefix: 'job-metadata/'
      }
    }
  });
  const s3Bucket = new Bucket(stack, 'TestM2C2Bucket');
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  const cloudWatchLogsPolicy = new Policy(stack, 'TestCloudWatchLogsPolicy');

  const jobBuilder = new JobBuilderConstruct(stack, 'TestJobBuilder', {
    cloudWatchLogsPolicy,
    greengrassGroupId: 'greengrass',
    kinesisStreamName: 'kinesis-stream',
    s3Bucket,
    s3LoggingBucket,
    sendAnonymousUsage: 'Yes',
    sourceCodeMapping,
    solutionMapping,
    uuid: 'uuid'
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(jobBuilder.getDynamoDbTableName()).toBeDefined();
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Environment: {
      Variables: {
        CONNECTOR_LAMBDA_ROLE: {
          'Fn::GetAtt': ['TestJobBuilderConnectorLambdaRole3498718A', 'Arn']
        },
        GREENGRASS_ID: 'greengrass',
        JOB_BUILDER_KEY: {
          'Fn::FindInMap': [ 'Solution', 'Parameters', 'JobMetadataPrefix' ]
        },
        JOB_DYNAMODB_TABLE: {
          Ref: 'TestJobBuilderDynamoTable68BFEB1E'
        },
        KINESIS_STREAM: 'kinesis-stream',
        SEND_ANONYMOUS_METRIC: 'Yes',
        SOLUTION_ID: {
          'Fn::FindInMap': [ 'Solution', 'Parameters', 'Id' ]
        },
        SOLUTION_VERSION:{
          'Fn::FindInMap': [ 'Solution', 'Parameters', 'Version' ]
        },
        SOURCE_S3_BUCKET:  {
          'Fn::Join': [
            '',
            [
              {
                'Fn::FindInMap':[ 'SourceCode', 'General', 'S3Bucket' ]
              },
              '-',
              {
                Ref: 'AWS::Region'
              }
            ]
          ]
        },
        SOURCE_S3_PREFIX: {
          'Fn::FindInMap': [ 'SourceCode', 'General', 'KeyPrefix' ]
        },
        UUID: 'uuid'
      }
    }
  });
  expect(stack).toHaveResourceLike('AWS::DynamoDB::Table', {
    KeySchema: [
      {
        AttributeName: 'jobid',
        KeyType: 'HASH'
      },
      {
        AttributeName: 'version',
        KeyType: 'RANGE'
      }
    ]
  });
});
