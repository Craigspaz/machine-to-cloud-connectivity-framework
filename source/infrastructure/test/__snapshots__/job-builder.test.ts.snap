// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 job builder test 1`] = `
Object {
  "Mappings": Object {
    "Solution": Object {
      "Parameters": Object {
        "Id": "SolutionId",
        "JobMetadataPrefix": "job-metadata/",
        "Version": "SolutionVersion",
      },
    },
    "SourceCode": Object {
      "General": Object {
        "KeyPrefix": "Prefix",
        "S3Bucket": "Bucket",
      },
    },
  },
  "Resources": Object {
    "TestJobBuilderConnectorLambdaRole3498718A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
      },
      "Type": "AWS::IAM::Role",
    },
    "TestJobBuilderDynamoTable68BFEB1E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "jobid",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "version",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "jobid",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "version",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TestJobBuilderJobBuilderPolicyD287C026": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The * resource is needed to control Greengrass resources.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":s3:::",
                    Object {
                      "Fn::FindInMap": Array [
                        "SourceCode",
                        "General",
                        "S3Bucket",
                      ],
                    },
                    "-",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    "/",
                    Object {
                      "Fn::FindInMap": Array [
                        "SourceCode",
                        "General",
                        "KeyPrefix",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestM2C2BucketBCB85E3A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestM2C2BucketBCB85E3A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "iot:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/m2c2/job/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "lambda:CreateFunction",
                "lambda:DeleteFunction",
                "lambda:PublishVersion",
                "lambda:CreateAlias",
                "lambda:DeleteAlias",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:m2c2-opcda-connector*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:m2c2-slmp-connector*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestJobBuilderConnectorLambdaRole3498718A",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "greengrass:CreateCoreDefinitionVersion",
                "greengrass:CreateCoreDefinition",
                "greengrass:CreateDeployment",
                "greengrass:CreateDeviceDefinitionVersion",
                "greengrass:CreateDeviceDefinition",
                "greengrass:CreateFunctionDefinitionVersion",
                "greengrass:CreateFunctionDefinition",
                "greengrass:CreateGroupVersion",
                "greengrass:CreateLoggerDefinitionVersion",
                "greengrass:CreateLoggerDefinition",
                "greengrass:CreateResourceDefinitionVersion",
                "greengrass:CreateResourceDefinition",
                "greengrass:CreateSubscriptionDefinitionVersion",
                "greengrass:CreateSubscriptionDefinition",
                "greengrass:DeleteCoreDefinition",
                "greengrass:DeleteDeviceDefinition",
                "greengrass:DeleteFunctionDefinition",
                "greengrass:DeleteLoggerDefinition",
                "greengrass:DeleteResourceDefinition",
                "greengrass:DeleteSubscriptionDefinition",
                "greengrass:GetCoreDefinitionVersion",
                "greengrass:GetDeploymentStatus",
                "greengrass:GetDeviceDefinitionVersion",
                "greengrass:GetFunctionDefinitionVersion",
                "greengrass:GetGroupVersion",
                "greengrass:GetGroup",
                "greengrass:GetLoggerDefinitionVersion",
                "greengrass:GetResourceDefinitionVersion",
                "greengrass:GetSubscriptionDefinitionVersion",
                "greengrass:ListCoreDefinitions",
                "greengrass:ListDeviceDefinitions",
                "greengrass:ListFunctionDefinitions",
                "greengrass:ListLoggerDefinitions",
                "greengrass:ListResourceDefinitions",
                "greengrass:ListSubscriptionDefinitions",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestJobBuilderJobBuilderPolicyD287C026",
        "Roles": Array [
          Object {
            "Ref": "TestJobBuilderLambdaFunctionServiceRoleD1AEB7EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestJobBuilderJobGeneratorRule3CD2931F": Object {
      "Properties": Object {
        "TopicRulePayload": Object {
          "Actions": Array [
            Object {
              "Lambda": Object {
                "FunctionArn": Object {
                  "Fn::GetAtt": Array [
                    "TestJobBuilderLambdaFunction199E5076",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Description": "Processing of the job submitted and generating the job file to be pushed to Greengrass",
          "RuleDisabled": false,
          "Sql": "SELECT * FROM 'm2c2/job/request'",
        },
      },
      "Type": "AWS::IoT::TopicRule",
    },
    "TestJobBuilderLambdaFunction199E5076": Object {
      "DependsOn": Array [
        "TestJobBuilderLambdaFunctionServiceRoleDefaultPolicyEF1113F8",
        "TestJobBuilderLambdaFunctionServiceRoleD1AEB7EC",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::FindInMap": Array [
                    "SourceCode",
                    "General",
                    "S3Bucket",
                  ],
                },
                "-",
                Object {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::FindInMap": Array [
                    "SourceCode",
                    "General",
                    "KeyPrefix",
                  ],
                },
                "/m2c2-job-builder.zip",
              ],
            ],
          },
        },
        "Description": "Machine to Cloud Connectivity job builder function",
        "Environment": Object {
          "Variables": Object {
            "CONNECTOR_LAMBDA_ROLE": Object {
              "Fn::GetAtt": Array [
                "TestJobBuilderConnectorLambdaRole3498718A",
                "Arn",
              ],
            },
            "GREENGRASS_ID": "greengrass",
            "JOB_BUILDER_BUCKET": Object {
              "Ref": "TestM2C2BucketBCB85E3A",
            },
            "JOB_BUILDER_KEY": Object {
              "Fn::FindInMap": Array [
                "Solution",
                "Parameters",
                "JobMetadataPrefix",
              ],
            },
            "JOB_DYNAMODB_TABLE": Object {
              "Ref": "TestJobBuilderDynamoTable68BFEB1E",
            },
            "KINESIS_STREAM": "kinesis-stream",
            "SEND_ANONYMOUS_METRIC": "Yes",
            "SOLUTION_ID": Object {
              "Fn::FindInMap": Array [
                "Solution",
                "Parameters",
                "Id",
              ],
            },
            "SOLUTION_VERSION": Object {
              "Fn::FindInMap": Array [
                "Solution",
                "Parameters",
                "Version",
              ],
            },
            "SOURCE_S3_BUCKET": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::FindInMap": Array [
                      "SourceCode",
                      "General",
                      "S3Bucket",
                    ],
                  },
                  "-",
                  Object {
                    "Ref": "AWS::Region",
                  },
                ],
              ],
            },
            "SOURCE_S3_PREFIX": Object {
              "Fn::FindInMap": Array [
                "SourceCode",
                "General",
                "KeyPrefix",
              ],
            },
            "UUID": "uuid",
          },
        },
        "Handler": "m2c2_job_builder_main.lambda_handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestJobBuilderLambdaFunctionServiceRoleD1AEB7EC",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 300,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestJobBuilderLambdaFunctionLambdaInvokePermission25B64893": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TestJobBuilderLambdaFunction199E5076",
            "Arn",
          ],
        },
        "Principal": "iot.amazonaws.com",
        "SourceAccount": Object {
          "Ref": "AWS::AccountId",
        },
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "TestJobBuilderJobGeneratorRule3CD2931F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestJobBuilderLambdaFunctionServiceRoleD1AEB7EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestJobBuilderLambdaFunctionServiceRoleDefaultPolicyEF1113F8": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray and access ENIs in a VPC.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestJobBuilderDynamoTable68BFEB1E",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestJobBuilderLambdaFunctionServiceRoleDefaultPolicyEF1113F8",
        "Roles": Array [
          Object {
            "Ref": "TestJobBuilderLambdaFunctionServiceRoleD1AEB7EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestLoggingBucket077C0B0B": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestM2C2BucketBCB85E3A": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
